market_analysis/views/default.py

```diff
@@ -0,0 +1,450 @@
+    search_results = []
+    if request.method == 'GET':
+        return {'stocks': search_results, 'msg': msg}
+    elif request.method == 'POST':
+        try:
+            search = request.params.get('search')
+            search1 = search.lower().capitalize()
+            search2 = search.upper()
+            search_query = request.dbsession.query(Stocks)\
+                .filter(or_(Stocks.name.startswith(search1),
+                        Stocks.symbol.startswith(search2)))
+        except DBAPIError:  # pragma: no cover
+            return Response(db_err_msg, content_type='text/plain', status=500)
+
+        for row in search_query:
+            search_results.append(row)
+        if len(search_results) == 0:
+            msg = 'No results found, try again.'
+        return {'stocks': search_results, 'msg': msg}
+
+
+@view_config(route_name='add', renderer='../templates/add_page.jinja2',
+             permission='secret')
+def add_stock_to_portfolio(request):
+    if request.method == 'POST':
+        current_user_id = request.dbsession.query(Users).filter(
+                Users.username == request.authenticated_userid
+            ).first().id
+        new_stock_id = request.matchdict['id']
+        query = request.dbsession.query(Association)\
+            .filter(Association.user_id == current_user_id)
+        list_of_stock_ids = []
+        for row in query:
+            list_of_stock_ids.append(row.stock_id)
+        if int(new_stock_id) not in list_of_stock_ids:
+            association_row = Association(user_id=current_user_id,
+                                          stock_id=new_stock_id, shares=0)
+            request.dbsession.add(association_row)
+            msg = request.matchdict['name'] + ' was added to your portfolio.'
+        else:
+            msg = request.matchdict['name'] + ' is already in your portfolio.'
+        return {'msg': msg}
+
+
+@view_config(route_name='delete', renderer='../templates/delete_page.jinja2',
+             permission='secret')
+def delete_stock_from_portfolio(request):
+    if request.method == 'POST':
+        current_user_id = request.dbsession.query(Users).filter(
+                Users.username == request.authenticated_userid
+            ).first().id
+        new_stock_sym = request.matchdict['sym']
+        try:
+            query = request.dbsession.query(Stocks).\
+             filter(Stocks.symbol == new_stock_sym).first()
+            query_del = request.dbsession.query(Association)\
+                .filter(and_(Association.stock_id == query.id,
+                             Association.user_id == current_user_id)).first()
+            request.dbsession.delete(query_del)
+            msg = request.matchdict['sym'] + ' was removed from'\
+                ' your portfolio.'
+
+        except (AttributeError, UnmappedInstanceError):
+            msg = 'Failed: tried to remove a stock that is not in the'\
+                ' portfolio.'
```

Comment:
As above, you could put this logic onto the `Users` model, making your views a bit thinner and easier to read.